### ISSUE TITLE
Throw exceptions when function/coroutine fails?

### ISSUE THREAD

**Issue by YukinoHayakawa** [ThePhD/sol2#841]
I'm using default error handlers set by `set_default_state` with `#define SOL_ALL_SAFETIES_ON 1`. I discovered that there are two suits of error handling and here is my understanding:

One is `on_error` parameter in `safe_script` which defaults to `script_default_on_error`.

Another is `default_traceback_error_handler` set by `protected_function::set_default_handler`.

Both of them check the return values from lua and invoke the corresponding handler if an error was found. `protected_function::invoke()` invokes `lua_pcall()` with the error handler so it get called if any error occurs. `safe_script()` checks the returned `protected_function_result` and throws (I have exceptions enabled) if the result isn't valid, and the result actually comes from `protected_function::invoke()`. The effect is `default_traceback_error_handler` will get called prior to `script_default_on_error` if error occurs during `safe_script()`. `coroutine` has similar behavior to `protected_function`.

I can see that functions and coroutines are designed to be on a lower level than scripts and exceptions are only thrown when executing scripts. What confuses me is, I did enable exceptions in the config so I expected exceptions to be thrown ASAP, but it actually silently fails to do so for functions and coroutines which are not less common than scripts. To get the default throwing behavior for them, I have to manually check the result and invoke `script_default_on_error`, and this duty is overlapped with `safe_script`. 

To make a guess, I think it is possible that this two-level error handling is also why #839 happens to avoid printing error messages twice. The error message is built by `default_traceback_error_handler` and thrown by `script_default_on_error`. But this does not happen when dealing with functions directly.

I think there could be more coherent way of error-handling and throwing exceptions. Maybe move the throws to a lower level or add new functions to check the invoke results and throws if errors occurred. But I'm not sure whether they are good ideas since I'm new to this library.

**Comment by ThePhD** [comment]
Throwing -- in general -- is actually really bad.

Exceptions destroy C APIs, because C APIs have no idea how to handle exceptions safely. We throw in the `panic` call because at that point the VM has signaled it is in a state of no return: throwing is a courtesy so that you can log what happens and bail, because at that point your VM is toast.

This changes when you use a later version of LuaJIT (on either x64 or use it on x86 (32-bit) with the LuaJIT beta versions), or when you compile Lua as C++. But note that even if you throw an exception on the first error and there are built-in ways to handle it, vanilla Lua compiled as C++ will still have no idea WTF to do with a thrown error and will essentially go into undefined behavior (read: crash happy, bug-prone) mode.

If you need the safety, `protected_function` has an `error_handler` member function. With all safeties defined, `sol::function` defaults to being `sol::protected_function`. You can also set the default error handler for _all_ protected functions to throw, but is it highly not advisable to do so as you may _still_ end up blowing up your stack or corrupting something.

Long story short: it's really bad. There's a small writeup in the docs on both exceptions and error handling:
https://sol2.readthedocs.io/en/latest/exceptions.html
https://sol2.readthedocs.io/en/latest/errors.html#catch-and-crash

It's painful to do manual error handling, yeah, but most other things result in kind of bad behavior. Not sure I can do much else for you other than tell you to change the defaults: https://sol2.readthedocs.io/en/latest/api/protected_function.html#id3

**Comment by YukinoHayakawa** [comment]
Thank you very much for your quick and informative reply.

I understand that throwing while having Lua C functions on the stack, like inside the error handler, is not good. I don't want to blow my stack up. What I mean is, why can't we have throwing functions similar to `safe_script` but for functions and coroutines which throws when an error was detected _after returned from lua_?

**Comment by ThePhD** [comment]
Because `sol::(safe_/protected_)function` and `sol::coroutine` are vocabulary types at a level too low for me to assume that is what the user wants as a default under any circumstances. Users already complained that triggering panic from `my_table["doesnt_exist"]["foo"]` and throwing provoked throw-based crashes that were exceptionally hard to recover from; I can't have `safe_function` or `safe_coroutine` doing that.

If you need it, there's no harm in writing your own `ns::throw_function f = lua["f"];` whose only job is to wrap `sol::protected_function` and throw on error. If I find a name for it in sol2 that isn't terrible, I might write one for `sol`.

Make sure you clean up the stack before throwing otherwise results might be left on the stack.

**Comment by YukinoHayakawa** [comment]
I see. Thank you.

**Comment by ThePhD** [comment]
(I'm very serious about finding a name for it in sol2: if we can find something that fits into the same space that makes it clear "this is a high level abstraction that is less configurable than the other stuff", then I'll gladly put it into sol2.)

**Comment by YukinoHayakawa** [comment]
Really appreciate that you are treating every user request very seriously. I'm always very glad to use libraries from great developers!
